secrets/config.go

```diff
@@ -0,0 +1,263 @@
+	data := []Config[map[string]any]{}
+	if err := unmarshal(&data); err != nil {
+		return fmt.Errorf("unable to unmarshal secret configurations: %w", err)
+	}
+	m := map[string]*ProviderConfig{}
+	for _, configMap := range data {
+		if secretProviderCount := len(configMap.Data); secretProviderCount == 0 {
+			return fmt.Errorf("expected secret provider but found none")
+		} else if secretProviderCount > 1 {
+			return fmt.Errorf("expected single secret provider but found %d", len(configMap.Data))
+		}
+		// For loop, but only 1 entry ever.
+		for name := range configMap.Data {
+			node := &yaml.Node{}
+			if err := node.Encode(configMap.Data[name]); err != nil {
+				return fmt.Errorf("unable to decode secret configuration: %w", err)
+			}
+
+			// Since we decode from any, we could potentially have a
+			// non-deterministic map.
+			nodeMappingSort(node)
+
+			providerList, ok := m[name]
+			if !ok {
+				*c = append(*c, ProviderConfig{
+					Name:    name,
+					Configs: []Config[*yaml.Node]{},
+				})
+				// Get reference to latest entry.
+				providerList = &((*c)[len(*c)-1])
+				m[name] = providerList
+			}
+			providerList.Configs = append(providerList.Configs, Config[*yaml.Node]{
+				Name: configMap.Name,
+				Data: node,
+			})
+		}
+	}
+	return nil
+}
+
+// MarshalYAML implements yaml.Marshaler.
+func (c *Configs) MarshalYAML() (interface{}, error) {
+	// Must marshal as any because yaml.v3.Node doesn't work in yaml.v2.
+	data := []Config[map[string]any]{}
+	for _, providerConfig := range *c {
+		for _, secretConfig := range providerConfig.Configs {
+			if secretConfig.Data.IsZero() {
+				continue
+			}
+			var secretData any
+			if err := secretConfig.Data.Decode(&secretData); err != nil {
+				return nil, err
+			}
+			data = append(data, Config[map[string]any]{
+				Name: secretConfig.Name,
+				Data: map[string]any{
+					providerConfig.Name: secretData,
+				},
+			})
+		}
+	}
+	return data, nil
+}
+
+// nodeMappingSort ensures mapping nodes have sorted key/value pair
+// ordering for deterministic YAML output.
+// TODO(TheSpiritXIII): This is no longer necessary with yaml.v3.
+func nodeMappingSort(node *yaml.Node) {
+	if node.Kind == yaml.MappingNode {
+		var keys []string
+		valueNodes := map[string]*yaml.Node{}
+		keyNodes := map[string]*yaml.Node{}
+		for i := 0; i < len(node.Content); i += 2 {
+			key := node.Content[i]
+			keyNodes[key.Value] = key
+			valueNodes[key.Value] = node.Content[i+1]
+			keys = append(keys, key.Value)
+		}
+
+		var content []*yaml.Node
+		slices.Sort(keys)
+		for _, key := range keys {
+			content = append(content, keyNodes[key], valueNodes[key])
+		}
+		node.Content = content
+	}
+}
+
+// ConfigsYAMLv3 is Configs but with YAML v3 unmarshalling. For custom
+// marshalling and unmarshalling, use a pointer to this type.
+type Configs2YAMLv3 []ProviderConfig
+
+// UnmarshalYAML implements yaml.v3.Unmarshaler.
+func (c *Configs2YAMLv3) UnmarshalYAML(node *yaml.Node) error {
```

Comment:
This is great, but I would add that once there is a need for yaml.v3 parsing of Prometheus config 🤔